File: web/packages/new/photos/services/comment.ts
import { encryptBox } from "ente-base/crypto";
import { authenticatedRequestHeaders, ensureOk } from "ente-base/http";
import { apiURL } from "ente-base/origins";
import { z } from "zod";

/**
 * A decrypted comment.
 */
export interface Comment {
    id: string;
    collectionID: number;
    fileID?: number;
    parentCommentID?: string;
    /** User ID of the parent comment's author (only in feed responses). */
    parentCommentUserID?: number;
    userID: number;
    anonUserID?: string;
    /** The decrypted comment text. */
    text: string;
    isDeleted: boolean;
    createdAt: number;
    updatedAt: number;
}

/**
 * Add a comment to a file in a collection.
 *
 * @param collectionID The ID of the collection containing the file.
 * @param fileID The ID of the file to comment on.
 * @param text The comment text to encrypt.
 * @param collectionKey The decrypted collection key (base64 encoded).
 * @param parentCommentID Optional parent comment ID for replies.
 * @returns The ID of the created comment.
 */
export const addComment = async (
    collectionID: number,
    fileID: number,
    text: string,
    collectionKey: string,
    parentCommentID?: string,
): Promise<string> => {
    const { encryptedData: cipher, nonce } = await encryptBox(
        new TextEncoder().encode(text),
        collectionKey,
    );

    const res = await fetch(await apiURL("/comments"), {
        method: "POST",
        headers: {
            ...(await authenticatedRequestHeaders()),
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            collectionID,
            fileID,
            cipher,
            nonce,
            parentCommentID,
        }),
    });
    ensureOk(res);
    const { id } = CreateCommentResponse.parse(await res.json());
    return id;
};

/**
 * Delete a comment by its ID.
 *
 * @param commentID The ID of the comment to delete.
 */
export const deleteComment = async (commentID: string): Promise<void> => {
    const res = await fetch(await apiURL(`/comments/${commentID}`), {
        method: "DELETE",
        headers: await authenticatedRequestHeaders(),
    });
    ensureOk(res);
};

const CreateCommentResponse = z.object({ id: z.string() });

File: mobile/apps/photos/lib/models/social/comment.dart
class Comment {
  final String id;
  final int collectionID;
  final int? fileID;
  final String data;
  final String? parentCommentID;
  final int? parentCommentUserID;
  final bool isDeleted;
  final int userID;
  final String? anonUserID;
  final int createdAt;
  final int updatedAt;

  Comment({
    required this.id,
    required this.collectionID,
    this.fileID,
    required this.data,
    this.parentCommentID,
    this.parentCommentUserID,
    this.isDeleted = false,
    required this.userID,
    this.anonUserID,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isAnonymous => userID <= 0;

  bool get isReply => parentCommentID != null;

  bool get isOnFile => fileID != null;

  Comment copyWith({
    String? id,
    int? collectionID,
    int? fileID,
    String? data,
    String? parentCommentID,
    int? parentCommentUserID,
    bool? isDeleted,
    int? userID,
    String? anonUserID,
    int? createdAt,
    int? updatedAt,
  }) {
    return Comment(
      id: id ?? this.id,
      collectionID: collectionID ?? this.collectionID,
      fileID: fileID ?? this.fileID,
      data: data ?? this.data,
      parentCommentID: parentCommentID ?? this.parentCommentID,
      parentCommentUserID: parentCommentUserID ?? this.parentCommentUserID,
      isDeleted: isDeleted ?? this.isDeleted,
      userID: userID ?? this.userID,
      anonUserID: anonUserID ?? this.anonUserID,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'collectionID': collectionID,
      'fileID': fileID,
      'data': data,
      'parentCommentID': parentCommentID,
      'parentCommentUserID': parentCommentUserID,
      'isDeleted': isDeleted,
      'userID': userID,
      'anonUserID': anonUserID,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }

  factory Comment.fromMap(Map<String, dynamic> map) {
    return Comment(
      id: map['id'] as String,
      collectionID: map['collectionID'] as int,
      fileID: map['fileID'] as int?,
      data: map['data'] as String,
      parentCommentID: map['parentCommentID'] as String?,
      parentCommentUserID: map['parentCommentUserID'] as int?,
      isDeleted: map['isDeleted'] as bool? ?? false,
      userID: map['userID'] as int,
      anonUserID: map['anonUserID'] as String?,
      createdAt: map['createdAt'] as int,
      updatedAt: map['updatedAt'] as int,
    );
  }
}
